addons/dynamishelper/dynamishelper.lua

```diff
@@ -24,408 +24,427 @@
-       				if item == nil then
-         				a,b,item = string.find(original,"%w+ obtains an? ..(%w+%-%w+ %w+)..\46")
-          					if item == nil then
-           						a,b,item = string.find(original,"%w+ obtains an? ..(%w+ %w+)..\46")
-         					end
-       				end
-     		end
--- 		a,b,item = string.find(original,"%w+ obtains an? ..(.*)..\46")
- 		if item ~= nil then
- 			item = item:lower()
- 			for i=1, #Currency do
-				if item == Currency[i]:lower() then
-   					Currency[Currency[i]] = Currency[Currency[i]] + 1
-   				end
- 			end
- 			obtainedf()
- 		end
-    	initializebox()
- 	end
- 	return new, color
-end)
-
-function obtainedf()
-	obtained = nil
- 	for i=1,#Currency do
- 		if Currency[Currency[i]] ~= 0 then
-  			if obtained == nil then
-  				obtained = " "
-  			end
-   			obtained = (obtained..Currency[i]..': '..Currency[Currency[i]]..' \n ')
- 		end
- 	end
-end
-
-windower.register_event('zone change', function(id)
-	goodzone = ProcZones:contains(id)
-	if not goodzone then
-		windower.text.set_visibility('proc_box', false)
-	end
+    local time = nil
+    local item = nil
+    local fiend = original:match("%w+'s attack staggers the (%w+)%!")
+    if settings.timer then
+        if fiend == 'fiend' then
+            windower.send_command('timers c "'..current_mob..'" 30 down stun')
+            return new, color
+        end
+    end
+    if string.find(original,"remaining in Dynamis.") then
+        obj_time = (tonumber(original:match("%d+")) * 60)
+        end_time = os.time() + obj_time
+    end
+    if string.find(original,"will be expelled from Dynamis") then
+        obj_time = (tonumber(original:match("%d+")) * 60)
+        end_time = os.time() + obj_time
+    end
+    if string.find(original,"Your stay in Dynamis has been extended by %d+ minutes.") then
+        end_time = end_time + (tonumber(original:match("%d+")) * 60)
+        w.time = os.date('!%H:%M:%S', end_time)
+    end
+    item = original:match("Obtained key item: ..(%w+ %w+ %w+ %w+)..\46")
+    if item ~= nil then
+        item = item:lower()
+        for granule in Granules:it() do
+            if item == granule:lower() then
+                w[granule] = 1
+                init_window()
+            end
+        end
+    end
+    item = original:match("%w+ obtains an? ..(%w+ %w+ %w+ %w+)..\46")
+    --a,b,item = string.find(original,"%w+ obtains an? ..(%w+ %w+ %w+ %w+)..\46")
+    if item == nil then
+        item = original:match("%w+ obtains an? ..(%w+ %w+ %w+)..\46")
+        --a,b,item = string.find(original,"%w+ obtains an? ..(%w+ %w+ %w+)..\46")
+        if item == nil then
+            item = original:match("%w+ obtains an? ..(%w+%-%w+ %w+)..\46")
+            --a,b,item = string.find(original,"%w+ obtains an? ..(%w+%-%w+ %w+)..\46")
+            if item == nil then
+                item = original:match("%w+ obtains an? ..(%w+ %w+)..\46")
+                --a,b,item = string.find(original,"%w+ obtains an? ..(%w+ %w+)..\46")
```

Comment:
While I like tight matching, this may be overdoing it :) If you want to catch all those individual cases you could as well just check for the individual items directly:
```lua
if original:match('%w+ obtains a 100 Byne Bill') or
   original:match('%w+ obtains an Ordelle Bronzepiece') or
   -- and so on
```

And in that case you might as well use your `Currencies` table:
```lua
if Currencies:any(function(currency) return original:match('%w+ obtains an? ..' .. currency .. '..\46') end) then
```

It does not get tighter than that. But will that not trigger for messages you receive for other characters as well? This is also not just your code, if it is an issue it was there before your refactoring as well. Or is that expected behavior?

This is honestly something that would be better solved using packets, but I guess that's outside of the scope of this refactoring.

Tl;dr I would either just match it more loosely as it was originally, or use exact matching using the table.

Reply:
Realistically, this plugin will be used for solo farming, so there's no risk of it matching another character in this situation.
However, nothing stops you from using it with a group (or while multi-boxing) and in both cases, tracking currencies just for you or for the group (so you have the total amount to split at the end) are both valid cases.
This could be made configurable, but I'm running out of time and will therefore change it to only get your own drops. Someone else can contribute later if they want it configurable.